### PR TITLE
Full rewrite of the crate

### LINKED ISSUE #12: Isn't the `LimineEntryPoint` type missing `extern "C"`?
Currently, the `LimineEntryPoint` type wraps an `fn() -> !`. Is there a reason why calling this function with the Rust calling convention works as expected? Shouldn't it be `extern "C" fn() -> !`?

Also, I've got a question:

The `LimineEntryPointRequest` structure stores a `LiminePtr<LimineEntryPoint>` internally. Is there a reason for that? I mean, `LiminePtr<LimineEntryPoint>` basically translates to `*mut fn() -> !`, and it pretty weird knowing that the "real" type is `fn() -> !`. In fact, the `LiminePtr::<LimineEntryPoint>::new` method creates a `LiminePtr` by casting the input `fn() -> !` into a `*mut fn() -> !`. Why not simply using `Option<LimineEntryPoint>` in `LimineEntryPointRequest`?

### LINKED ISSUE #13: `ArrayPtr::into_slice` and `ArrayPtr::into_slice_mut` should be `unsafe`
Those two functions are defined as follows:

```Rust
fn into_slice<'a>(&'a self, len: usize) -> &'a [NonNullPtr<T>] {
    // SAFETY: We have shared reference to the array.
    unsafe { core::slice::from_raw_parts(self.as_ptr(), len) }
}

fn into_slice_mut<'a>(&'a mut self, len: usize) -> &'a mut [NonNullPtr<T>] {
    // SAFETY: We have exculusive access to the array.
    unsafe { core::slice::from_raw_parts_mut(self.as_ptr(), len) }
}
```

But it is perfectly valid to call the function with a `len` that's way too large for the data actually referenced by the pointer. The pointer knows it references an *array*, but it does not know *how many* items it points to.

The safety doc of the functions should include something like this:

```Rust
/// # Safety
///
/// The specified `len` must not exceed the number of elements actually referenced by the pointer. 
```

... or something like that.

I'd be down to implement the pull request for this though.

### LINKED ISSUE #14: It's possible to alias mutable references to a single response through safe code
Currently, it is possible to alias mutable references to Limine responses in safe code.

Every request type returns a owned `LiminePtr` to their associated response. `LiminePtr`'s API provides a `get` and a `get_mut` method to access the underlying data. In other words, it is possible to call the `*Request::get_response` function to get two copies of the same `LiminePtr` on two separate threads, then use those two copies to create two separate exclusive references to the underlying response.

I'm pretty sure the intent of `LiminePtr` is to "own" the value without owning the place. The pointer is *itself* supposed to remain unique, so it seems weird to me that it is returned by the `get_response` method.

One way to fix that would be to avoid returning the underlying `LiminePtr` directly, and providing two methods to access the response:

```rust
impl *Request {
    pub fn response(&self) -> Option<&*Response> {
        // SAFETY:
        //  The `*Request` is borrowed by reference, and our API never gives out a
        //  mutable reference from a regular `*Request` reference, ensuring that
        //  this dereference is safe.
        unsafe { &*self.response.get() }.get()
    }

    pub fn response_mut(&mut self) -> Option<&mut *Response> {
        // SAFETY: same argument, but for the exclusivity of the pointer.
        unsafe { &mut *self.response.get() }.get_mut()
    }
}
```

... or we could simply use those methods as accessors to internal `LiminePtr`, but that would just make the API harder to use.

```rust
fn response(&self) -> &LiminePtr<*Response>;
fn response_mut(&mut self) -> &mut LiminePtr<*Resonse>;
```

With this new API, you need a mutable reference to the request in order to have a mutable reference to the response. And, in order to have a mutable reference to the request... Well.. You need a `static mut`, which is unsafe to use *because* it could be accessed from two separate threads.

Also, isn't `LiminePtr` just a `Option<NonNullPtr>`? Is there any reason to re-implement all that logic here? Why not define a type alias?

I'd be down to write a PR for this issue!

### LINKED ISSUE #16: The stored `LimineTerminalWrite` function can be used in invalid ways.
Currently, (and even when fixing #14), it is possible to call the `LimineTerminalWrite` function on two separate threads with the same `LimineTerminal` instance.

That one is pretty simple. The function returns by `LimineTerminalResponse::write` method returns a function that takes a shared `&LimineTerminal`, meaning that the function can be called simultaneously with the same aliased terminal.

Using the same exact trick, it's possible to call the write within itself in the terminal callback implementation.

Both of those operations are invalid (as specified in the specification).

An easy way to fix this would be to require the `write` function to take an exclusive reference to the target terminal.

Going from:

```rust
fn write(&self) -> Option<impl Fn(&Terminal, s: &str)>;
```

to:

```rust
fn write(&self) -> Option<impl Fn(&mut Terminal, s: &str)>;
```

Note that this would require an additional `terminals_mut` method to be added to the `LimineTerminalResponse`.

Also: is there any way to remove the `Option` here? I mean, does it even make sense for this function not to be present?

### PR DESCRIPTION
**Pull request by lylythechosenone**
This crate is not in great shape. It has [4 unresolved issues, all of which are UB](https://github.com/limine-bootloader/limine-rs/issues). It also does not correctly support platform-specific features. The API is confusing, and leaves much to be desired.

In this rewrite, I took into account my own struggles with the crate, as well as comments from other users. The history of this crate is rough, but I hope we can turn it around starting now.

I'll note that my changes are heavily opinionated (rustfmt, clippy with `-D warnings`, etc.), so feel free to bring up your concerns if you have any. I also have not implemented any tests, but I figure we can implement tests through the barebones project.

This fixes #12 and fixes #14, and indirectly fixes #13 and indirectly fixes #16 by way of removing them.

Edit: GitHub doesn't understand "and"

### CODE REVIEW

**Comment by lylythechosenone** [comment]
Alright, everything should be done now. Merge at your earliest convenience.

File: src/response.rs [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+    name: *const c_char,
+    version: *const c_char,
+}
+unsafe impl Sync for BootloaderInfoResponse {}
+unsafe impl Send for BootloaderInfoResponse {}
+impl BootloaderInfoResponse {
+    impl_base_fns!();
+
+    /// Returns the name of the loading bootloader.
+    pub fn name(&self) -> &str {
```
Comment (48cf) [inline src/response.rs:40]: Should these 2 functions perhaps return a `Result` instead of possibly panicking?

**Review by 48cf — CHANGES_REQUESTED** [review]
Please make up your mind about the spacing in between items and whether the doc comments should be put before or after attributes. The spacing is inconsistent with multiple items having no spacing in between and some doc comments being put between the attributes and the item they belong to

File: src/file.rs [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+#[repr(C)]
+pub struct File {
+    revision: u64,
+    addr: *mut c_void,
+    size: u64,
+    path: *const c_char,
+    cmdline: *const c_char,
+    media_type: MediaType,
+    _unused: MaybeUninit<u32>,
+    tftp_ip: Option<NonZeroU32>,
```
[the remark this answers is not in the document]
Comment (lylythechosenone) [reply src/file.rs:74]: Yes, it is.

File: src/file.rs [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        self.tftp_ip()
+            .and_then(|ip| Some(SocketAddrV4::new(ip, self.tftp_port()?.get() as u16)))
+    }
+
+    /// The partition index of the file, if the file was loaded from a partition.
+    pub fn partition_idx(&self) -> Option<NonZeroU32> {
+        self.partition_idx
+    }
+    /// The MBR disk ID of the file, if the file was loaded from an MBR disk.
+    pub fn mbr_disk_id(&self) -> u32 {
```
[the remark this answers is not in the document]
Comment (lylythechosenone) [reply src/file.rs:150]: Oops, didn't catch this one. Will do.

File: src/request.rs [hunk 4]
```diff
@@ -0,0 +1,786 @@
...
+            inner: UnsafeCell::new(None),
+        }
+    }
+}
+
+/// Request the name and version of the loading bootloader.
+///
+/// # Usage
+/// ```rust
+/// # use limine::{request::BootloaderInfoRequest, response::BootloaderInfoResponse, BaseRevision};
```
[the remark this answers is not in the document]
Comment (lylythechosenone) [reply src/request.rs:103]: This is a rustdoc feature that hides the line from the generated docs.

File: src/response.rs (hunk 1 above)
Comment (lylythechosenone) [reply src/response.rs:40]: According to the spec, they're ASCII strings. Therefore, they must be valid UTF-8.

File: src/file.rs [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+    }
+    /// The size of the file, in bytes.
+    pub fn size(&self) -> u64 {
+        self.size
+    }
+
+    /// The path of the file. This is the path that was passed to the bootloader
+    /// in either the configuration file or the `internal_modules` field of the
+    /// [`ModuleRequest`](crate::request::ModuleRequest).
+    pub fn path(&self) -> Result<&str, &[u8]> {
```
[the remark this answers is not in the document]
Comment (lylythechosenone) [reply src/file.rs:101]: Will add this to the doc.

**Comment by lylythechosenone** [comment]
Yeah sorry about that, I thought I caught all of those. I'll do a regex search I guess.

**Review by 48cf — APPROVED** [review]

**Pull request merged**